### ISSUE TITLE
Calling waitpid() on Windows can result in "panic: memory wrap"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16529]
Migrated from [rt.perl.org#133145](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133145) (status was 'new')

Searchable as RT133145$

**Comment by p5pRT** [comment]
### From eajspm@gmail.com

#### Created by eajspm@gmail.com

This is a bug report for perl from eajspm@&#8203;gmail\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.26\.1\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Calling waitpid\(\-1\, WNOHANG\) in a Perl script on Windows can result in the
Perl interpreter aborting with the error&#8203;:

panic&#8203;: memory wrap at script\.pl line xx

The bug appears to be in either win32\_internal\_wait\(\) or win32\_msgwait\(\) in
the
source file win32\\win32\.c\.  win32\_msgwait\(\) calls
MsgWaitForMultipleObjects\(\)\.
If that call returns WAIT\_OBJECT\_0 \+ count\, this is a special case that
means a
new message is available in the thread's input queue\.

When this happens\, the result WAIT\_OBJECT\_0 \+ count is returned to
win32\_internal\_wait\(\)\, but that function does not know how to properly deal
with that return code\.  It makes the bad assumption that the return code is
the
index of a thread that has finished and probably indexes into an invalid
location in an array or otherwise ends up corrupting memory somewhere \(I
didn't
track down the exact source of the panic\)\.

I put in a local fix to make win32\_internal\_wait\(\) return 0 when
win32\_msgwait\(\) returns WAIT\_OBJECT\_0 \_\+ count and it solved the problem in
my
test program\.  A better solution would probably be to have win32\_msgwait\(\)
call
MsgWaitForMultipleObjects\(\) again in this case since it technically didn't
complete the job it was asked to do\.  However\, this is the first I've ever
looked at the Perl source code so I'm just guessing here\.

There are several other places that call win32\_msgwait\(\)\.  They should all
be
inspected to see if they need to handle this return code as well\.

Below is a simple program that I used to reproduce the problem and verify my
simple fix\.  I can readily reproduce the problem on one particular Windows
machine\, but I cannot reproduce it on several other Windows machines\.  I'm
not
sure what the differences are\.  Perhaps it's because the machine where I can
reproduce it is a large machine with 16 cores and it can achieve a higher
level
of parallelism with the fork\(\) calls\.  Either way\, the bug is obvious if you
look through the source code\.

=================================================================
use strict;
use warnings;

use POSIX "&#8203;:sys\_wait\_h";

\# This is a simple program that spawns new "processes" with fork\(\) as fast
as
\# possible\.  The child processes do nothing except exit immediately\.  Each
time
\# we fork a new process\, we call waitpid\(\) to clean up any child processes
that
\# have finished\.  That way we should never have more than a few processes
\# outstanding\.
for \(my $i = 1; $i \<= 30000; $i\+\+\)
\{
&nbsp;    my $pid = fork\(\);
&nbsp;    if \(\! defined\($pid\)\)
&nbsp;    \{
&nbsp;        die "\*\*\* fork\(\) error&#8203;: $\!\\n";
&nbsp;    \}
&nbsp;    elsif \($pid\)
&nbsp;    \{
&nbsp;        \# Parent\.
&nbsp;        my $done = 0;
&nbsp;        while \(\! $done\)
&nbsp;        \{
&nbsp;            my $ret = waitpid\(\-1\, &WNOHANG\);
&nbsp;            if \($ret == 0\) \{
&nbsp;                \# No processes ready to reap\.
&nbsp;                $done = 1;
&nbsp;            \} elsif \($ret == \-1\) \{
&nbsp;                \# Error case\.  Not sure what we can do except retry later\.
&nbsp;                $done = 1;
&nbsp;            \} else \{
&nbsp;                \# Success case\.  Reaped a child process\.  Go around again\.
&nbsp;            \}
&nbsp;        \}
&nbsp;    \}
&nbsp;    else
&nbsp;    \{
&nbsp;        \# Child\.
&nbsp;        print "Hello from child $i\\n";
&nbsp;        exit;
&nbsp;    \}
\}
=================================================================

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.26.1:

Configured by strawberry-perl at Sat Sep 23 23:10:19 2017.

Summary of my perl5 (revision 5 version 26 subversion 1) configuration:

  Platform:
    osname=MSWin32
    osvers=6.3
    archname=MSWin32-x86-multi-thread-64int
    uname='Win32 strawberry-perl 5.26.1.1 #1 Sat Sep 23 23:07:28 2017 i386'
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='gcc'
    ccflags =' -s -O2 -DWIN32 -D__USE_MINGW_ANSI_STDIO
-DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS
-DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-s -O2'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='7.1.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=4
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=12
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++.exe'
    ldflags ='-s -L"D:\Apps\SPerl32Portable\perl\lib\CORE"
-L"D:\Apps\SPerl32Portable\c\lib"'
    libpth=D:\Apps\SPerl32Portable\c\lib
D:\Apps\SPerl32Portable\c\i686-w64-mingw32\lib
D:\Apps\SPerl32Portable\c\lib\gcc\i686-w64-mingw32\7.1.0
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=
    so=dll
    useshrplib=true
    libperl=libperl526.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=xs.dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-mdll -s -L"D:\Apps\SPerl32Portable\perl\lib\CORE"
-L"D:\Apps\SPerl32Portable\c\lib"'



@INC for perl 5.26.1:
    d:/Apps/SPerl32Portable/perl/site/lib
    d:/Apps/SPerl32Portable/perl/vendor/lib
    d:/Apps/SPerl32Portable/perl/lib


Environment for perl 5.26.1:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=d:\Apps\SPerl32Portable\c\i686-w64-mingw32\bin;d:\Apps\SPerl32Portable\c\bin;d:\Apps\SPerl32Portable\perl\site\bin;d:\Apps\SPerl32Portable\perl\bin;d:\rt\Apps\7-ZipPortable\App\7-Zip64;d:\rt\Apps\curl-7.57.0\bin;d:\rt\Apps\wget-1.19.1;C:\ProgramData\Oracle\Java\javapath;C:\Program
Files (x86)\NVIDIA
Corporation\PhysX\Common;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Users\Erik\AppData\Local\Microsoft\WindowsApps
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by xenu** [comment]
Fixed in 08e55ec5e3ef6d6c040c0dc8bdec7d59f76bfbe8.

**Comment by khwilliamson** [comment]
Reopening to get in 5.32.1

**Comment by khwilliamson** [comment]
This fix caused another bug fixed in 81295a42ad, and that should go into 5.30.1

**Comment by tonycoz** [comment]
Fixed in all supported releases.
